### PR TITLE
Outreachy pdf generator

### PR DESCRIPTION
**Pull request by kavindya89**
### Proposed changes in this pull request

This pull request include the changes which done as part of outreachy internship. The project idea is detailed in https://github.com/Cadasta/cadasta-project-ideas/issues/12

-

### When should this PR be merged

After reviewing the pull request and testing the PDF form generation functionality

### Risks

This introduces a new functionality.

-

### Follow up actions

[List any possible follow-up actions here; for instance, testing data
migrations, software that we need to install on staging and production
environments.]

-

#### General

- [x] **Is this PR explained thoroughly?** All code changes must be accounted for in the PR description. 
- [ ] **Is the PR labeled correctly?** It should have the `migration` label if a new migration is added. 
- [ ] **Is the risk level assessment sufficient?** The risks section should contain all risks that might be introduced with the PR and which actions we need to take to mitigate these risks. Possible risks are database migrations, new libraries that need to be installed or changes to deployment scripts.

#### Functionality

- [ ] **Are all requirements met?** Compare implemented functionality with the requirements specification.
- [x] **Does the UI work as expected?** There should be no Javascript errors in the console; all resources should load. There should be no unexpected errors. Deliberately try to break the feature to find out if there are corner cases that are not handled. 

#### Code

- [x] **Do you fully understand the introduced changes to the code?** If not ask for clarification, it might uncover ways to solve a problem in a more elegant and efficient way.
- [ ] **Does the PR introduce any inefficient database requests?** Use the debug server to check for duplicate requests. 
- [ ] **Are all necessary strings marked for translation?** All strings that are exposed to users via the UI must be [marked for translation](https://docs.djangoproject.com/en/1.10/topics/i18n/translation/). 

#### Tests

- [x] **Are there sufficient test cases?** Ensure that all components are tested individually; models, forms, and serializers should be tested in isolation even if a test for a view covers these components. 
- [ ] **If this is a new feature or a significant change to an existing feature** has the manual testing spreadsheet been updated with instructions for manual testing?

#### Documentation

- [ ] **Are changes to the UI documented in the platform docs?** If this PR introduces new platform site functionality or changes existing ones, the changes must be documented in the [Cadasta Platform Documentation](https://github.com/Cadasta/cadasta-docs).
- [ ] **Are changes to the API documented in the API docs?** If this PR introduces new API functionality or changes existing ones, the changes must be documented in the [API docs](https://github.com/Cadasta/api-docs).
- [ ] **Are reusable components documented?** If this PR introduces components that are relevant to other developers (for instance a mixin for a view or a generic form) they should be documented in the Wiki.

### CODE REVIEW

File: cadasta/questionnaires/download/generator.py [hunk 1]
```diff
@@ -0,0 +1,61 @@
+from questionnaires.models import Question, QuestionGroup
+from questionnaires.models import Questionnaire
+from django.template.loader import render_to_string
+from weasyprint import HTML, CSS
+import logging
```
Comment (bjohare) [inline cadasta/questionnaires/download/generator.py]: This can be removed.

File: cadasta/questionnaires/download/generator.py [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+
+
+class PDFGenerator():
+    def __init__(self, project, pdfform):
+        self.project = project
+        self.pdfform = pdfform
+
+    def generate_pdf(self, absolute_uri):
+        logging.getLogger('weasyprint').setLevel(100)
```
Comment (bjohare) [inline cadasta/questionnaires/download/generator.py]: This can be removed

File: cadasta/questionnaires/download/generator.py [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+        template_questions_list = []
+        template_question_groups_list = []
+        question_groups_list = []
+        pdf = None
+
+        questionnaire = Questionnaire.objects.get(
+            id=self.project.current_questionnaire)
+        questions_list = Question.objects.filter(
```
Comment (bjohare) [inline cadasta/questionnaires/download/generator.py]: use `question_list = questionnaire.questions`

File: cadasta/questionnaires/download/generator.py [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+            questionnaire=questionnaire)
+        question_groups_list = QuestionGroup.objects.filter(
```
Comment (bjohare) [inline cadasta/questionnaires/download/generator.py]: use `question_groups_list = questionnaire.question_groups`

File: cadasta/questionnaires/download/generator.py [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+            questionnaire=questionnaire)
+
+        for question in questions_list:
+            if question.question_group is None:
+                question_option_list = question.options.all()
```
Comment (bjohare) [inline cadasta/questionnaires/download/generator.py]: These two lines can be removed. The question already has the list of options via its `question.options` property.

File: cadasta/questionnaires/download/generator.py [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+                question.question_option_list = question_option_list
+                template_questions_list.append(question)
+
+        for question_group in question_groups_list:
+            questions_list = Question.objects.filter(
```
Comment (bjohare) [inline cadasta/questionnaires/download/generator.py]: use `question_list = question_group.questions` here instead..

File: cadasta/questionnaires/download/generator.py [hunk 7]
```diff
@@ -0,0 +1,61 @@
...
+                question_group=question_group)
+
+            if questions_list is not None and len(questions_list) > 0:
+                for question in questions_list:
+                    question_option_list = question.options.all()
```
Comment (bjohare) [inline cadasta/questionnaires/download/generator.py]: Same as above here, the question already knows it's options via the `question.options` property..

File: cadasta/questionnaires/download/generator.py [hunk 8]
```diff
@@ -0,0 +1,61 @@
...
+                    question.question_option_list = question_option_list
+                    question_group.questions_list = questions_list
```
Comment (bjohare) [inline cadasta/questionnaires/download/generator.py]: This is not necessary as `question_group` already has the list of questions via `question_group.questions`. Have a look at the ForeignKey relations between the various Questionnaire models in `questionnaires/models.py`. ForeignKey relations add properties to the foreign model using the `related_name` argument which means that additional DB queries to lookup the related models are not necessary.

File: cadasta/questionnaires/tests/test_views_default.py [hunk 9]
```diff
@@ -0,0 +1,537 @@
...
+        }
+
+    def test_get_with_authorized_user(self):
+        user = UserFactory.create()
+        assign_policies(user)
+        response = self.request(user=user)
+        assert response.status_code == 200
+        assert response.content == self.expected_content
+
+    def test_get_from_non_existend_project(self):
```
Comment (bjohare) [inline cadasta/questionnaires/tests/test_views_default.py]: typo.. should be `test_get_from_non_existent_project`

File: cadasta/questionnaires/models.py [hunk 10]
```diff
@@ -198,3 +197,52 @@ def __repr__(self):
...
+
+    class Meta:
+        ordering = ('name',)
+
+    class TutelaryMeta:
+        perm_type = 'pdfform'
+        path_fields = ('project', 'pk')
+        actions = (
+            ('pdfform.add',
+             {'description': _("Add a pdf form of questionnaire "
```
Comment (bjohare) [inline cadasta/questionnaires/models.py]: Change to `Add a pdf form to the project`

File: cadasta/questionnaires/models.py [hunk 11]
```diff
@@ -198,3 +197,52 @@ def __repr__(self):
...
+                               "to the project"),
+              'error_message': messages.QUESTIONNAIRE_PDF_FORM_ADD,
+              'permissions_object': 'project'}),
+            ('pdfform.edit',
+             {'description': _("Edit a pdf form of questionnaire "
```
Comment (bjohare) [inline cadasta/questionnaires/models.py]: change to `Edit a pdf form`

File: cadasta/questionnaires/models.py [hunk 12]
```diff
@@ -198,3 +197,52 @@ def __repr__(self):
...
+                               "to the project"),
+              'error_message': messages.QUESTIONNAIRE_PDF_FORM_EDIT}),
+            ('pdfform.view',
+             {'description': _("View a pdf form of questionnaire "
```
Comment (bjohare) [inline cadasta/questionnaires/models.py]: change to `View a pdf form`

File: cadasta/questionnaires/models.py [hunk 13]
```diff
@@ -198,3 +197,52 @@ def __repr__(self):
...
+                               "to the project"),
+              'error_message': messages.QUESTIONNAIRE_PDF_FORM_VIEW}),
+            ('pdfform.list',
+             {'description': _("List pdf forms of questionnaire "
```
Comment (bjohare) [inline cadasta/questionnaires/models.py]: change to `List pdf forms`

File: cadasta/questionnaires/models.py [hunk 14]
```diff
@@ -198,3 +197,52 @@ def __repr__(self):
...
+                               "to the project"),
+              'error_message': messages.QUESTIONNAIRE_PDF_FORM_VIEW,
+              'permissions_object': 'project'}),
+            ('pdfform.delete',
+             {'description': _("Delete a pdf form of questionnaire "
```
Comment (bjohare) [inline cadasta/questionnaires/models.py]: change to `Delete a pdf form`

File: cadasta/questionnaires/models.py [hunk 15]
```diff
@@ -198,3 +197,52 @@ def __repr__(self):
...
+                               "to the project"),
+              'error_message': messages.QUESTIONNAIRE_PDF_FORM_DELETE}),
+            ('pdfform.generate',
+             {'description': _("Generate a pdf form of questionnaire "
```
Comment (bjohare) [inline cadasta/questionnaires/models.py]: change to `Generate a pdf form`

File: cadasta/questionnaires/messages.py [hunk 16]
```diff
@@ -6,3 +6,16 @@
                       "this project")
 QUESTIONNAIRE_EDIT = _("You don't have permission to edit the "
                        "questionnaire of this project")
+QUESTIONNAIRE_PDF_FORM_ADD = _("You don't have permission to add pdf form "
```
Comment (bjohare) [inline cadasta/questionnaires/messages.py]: change to `You don't have permission to add a pdf form to this project`

File: cadasta/questionnaires/messages.py [hunk 17]
```diff
@@ -6,3 +6,16 @@
...
+                               "of the questionnaire to this project")
+QUESTIONNAIRE_PDF_FORM_EDIT = _("You don't have permission to edit pdf form "
```
Comment (bjohare) [inline cadasta/questionnaires/messages.py]: change to `You don't have permission to edit this pdf form`

File: cadasta/questionnaires/messages.py [hunk 18]
```diff
@@ -6,3 +6,16 @@
...
+                                "of the questionnaire to this project")
+QUESTIONNAIRE_PDF_FORM_VIEW = _("You don't have permission to view pdf "
```
Comment (bjohare) [inline cadasta/questionnaires/messages.py]: change to `You don't have permission to view this pdf form`

File: cadasta/questionnaires/messages.py [hunk 19]
```diff
@@ -6,3 +6,16 @@
...
+                                "form of the questionnaire to this project")
+QUESTIONNAIRE_PDF_FORM_DELETE = _("You don't have permission to delete pdf "
```
Comment (bjohare) [inline cadasta/questionnaires/messages.py]: change to `You don't have permission to delete this pdf form`

File: cadasta/questionnaires/messages.py [hunk 20]
```diff
@@ -6,3 +6,16 @@
...
+                                  "form of the questionnaire to this project")
+QUESTIONNAIRE_PDF_FORM_LIST = _("You don't have permission to list pdf forms "
```
Comment (bjohare) [inline cadasta/questionnaires/messages.py]: change to `You don't have permission to list pdf forms for this project`

File: cadasta/questionnaires/messages.py [hunk 21]
```diff
@@ -6,3 +6,16 @@
...
+                                "of the questionnaire to this project")
+QUESTIONNAIRE_PDF_FORM_GENERATE = _("You don't have permission to generate "
```
Comment (bjohare) [inline cadasta/questionnaires/messages.py]: change to `You don't have permission to generate pdf forms for this project`

File: cadasta/questionnaires/tests/test_views_default.py [hunk 22]
```diff
@@ -0,0 +1,537 @@
...
+        response = self.request()
+        assert response.status_code == 302
+        assert '/account/login/' in response.location
+
+    def test_post_with_authorized_user(self):
+        user = UserFactory.create()
+        assign_policies(user)
+        response = self.request(method='POST', user=user)
+        assert PDFForm.objects.count() == 1
+        self.pdfform_created = PDFForm.objects.first()
```
Comment (bjohare) [inline cadasta/questionnaires/tests/test_views_default.py]: no need to assign this as an instance variable here `pdfform_created = PDFForm.objects.first()` is enough. You should test the properties of `pdfform_created`, eg assert that `pdfform_created.name == 'Some name'` etc..

File: cadasta/questionnaires/tests/test_views_default.py [hunk 23]
```diff
@@ -0,0 +1,537 @@
...
+        assert response.status_code == 302
+        assert response.location == self.expected_success_url
+
+    def test_post_with_unauthorized_user(self):
+        user = UserFactory.create()
+        response = self.request(method='POST', user=user)
+        assert PDFForm.objects.count() == 0
+        assert response.status_code == 302
+        assert ("You don't have permission to add pdf form of the "
```
Comment (bjohare) [inline cadasta/questionnaires/tests/test_views_default.py]: change to `You don't have permission to add a pdf form to this project`

File: cadasta/questionnaires/tests/test_views_default.py [hunk 24]
```diff
@@ -0,0 +1,537 @@
...
+        }
+
+    def test_get_with_authorized_user(self):
+        user = UserFactory.create()
+        assign_policies(user)
+        response = self.request(user=user)
+        assert response.status_code == 200
+        assert response.content == self.expected_content
+
+    def test_get_from_non_existend_project(self):
```
Comment (bjohare) [inline cadasta/questionnaires/tests/test_views_default.py]: change to `test_get_with_non_existent_project`

File: cadasta/questionnaires/tests/test_views_default.py [hunk 25]
```diff
@@ -0,0 +1,537 @@
...
+        user = UserFactory.create()
+        assign_policies(user)
+        with pytest.raises(Http404):
+            self.request(user=user, url_kwargs={'project': 'abc123'})
+
+    def test_get_non_existend_form(self):
```
Comment (bjohare) [inline cadasta/questionnaires/tests/test_views_default.py]: change to `test_get_non_existent_form`

File: cadasta/questionnaires/tests/test_views_default.py [hunk 26]
```diff
@@ -0,0 +1,537 @@
...
+        user = UserFactory.create()
+        assign_policies(user)
+        with pytest.raises(Http404):
+            self.request(user=user, url_kwargs={'form': 'abc123'})
+
+    def test_get_with_unauthorized_user(self):
+        user = UserFactory.create()
+        response = self.request(user=user)
+        assert response.status_code == 302
+        assert ("You don't have permission to view pdf form "
```
Comment (bjohare) [inline cadasta/questionnaires/tests/test_views_default.py]: should be changed to `You don't have permission to view this pdf form`

File: cadasta/questionnaires/tests/test_views_default.py [hunk 27]
```diff
@@ -0,0 +1,537 @@
...
+        assign_policies(user)
+        response = self.request(user=user)
+        assert response.status_code == 200
+        assert response.content == self.expected_content
+
+    def test_get_with_unauthorized_user(self):
+        user = UserFactory.create()
+        response = self.request(user=user)
+        assert response.status_code == 302
+        assert ("You don't have permission to delete pdf form "
```
Comment (bjohare) [inline cadasta/questionnaires/tests/test_views_default.py]: `You don't have permission to delete this pdf form`

File: cadasta/questionnaires/tests/test_views_default.py [hunk 28]
```diff
@@ -0,0 +1,537 @@
...
+                "of the questionnaire to this project"
+                in response.messages)
+
+    def test_get_with_unauthenticated_user(self):
+        response = self.request()
+        assert response.status_code == 302
+        assert '/account/login/' in response.location
+
+    def test_get_from_non_existend_project(self):
```
Comment (bjohare) [inline cadasta/questionnaires/tests/test_views_default.py]: `test_get_from_non_existent_project`

File: cadasta/questionnaires/tests/test_views_default.py [hunk 29]
```diff
@@ -0,0 +1,537 @@
...
+        user = UserFactory.create()
+        assign_policies(user)
+        with pytest.raises(Http404):
+            self.request(user=user, url_kwargs={'project': 'abc123'})
+
+    def test_get_non_existend_pdfform(self):
```
Comment (bjohare) [inline cadasta/questionnaires/tests/test_views_default.py]: `test_get_non_existent_project`

File: cadasta/questionnaires/tests/test_views_default.py [hunk 30]
```diff
@@ -0,0 +1,537 @@
...
+        assert response.status_code == 302
+        assert response.location == self.expected_success_url
+
+        assert PDFForm.objects.count() == 0
+
+    def test_post_with_unauthorized_user(self):
+        user = UserFactory.create()
+        response = self.request(method='POST', user=user)
+        assert response.status_code == 302
+        assert ("You don't have permission to delete pdf form "
```
Comment (bjohare) [inline cadasta/questionnaires/tests/test_views_default.py]: `You don't have permission to delete this pdf form`

File: cadasta/questionnaires/tests/test_views_default.py [hunk 31]
```diff
@@ -0,0 +1,537 @@
...
+        assign_policies(user)
+        response = self.request(user=user)
+        assert response.status_code == 200
+        assert response.content == self.expected_content
+
+    def test_get_with_unauthorized_user(self):
+        user = UserFactory.create()
+        response = self.request(user=user)
+        assert response.status_code == 302
+        assert ("You don't have permission to edit pdf form "
```
Comment (bjohare) [inline cadasta/questionnaires/tests/test_views_default.py]: `You don't have permission to edit this pdf form`

File: cadasta/questionnaires/tests/test_views_default.py [hunk 32]
```diff
@@ -0,0 +1,537 @@
...
+                "of the questionnaire to this project"
+                in response.messages)
+
+    def test_get_with_unauthenticated_user(self):
+        response = self.request()
+        assert response.status_code == 302
+        assert '/account/login/' in response.location
+
+    def test_get_from_non_existend_project(self):
```
Comment (bjohare) [inline cadasta/questionnaires/tests/test_views_default.py]: `test_get_from_non_existent_project`

File: cadasta/questionnaires/tests/test_views_default.py [hunk 33]
```diff
@@ -0,0 +1,537 @@
...
+        user = UserFactory.create()
+        assign_policies(user)
+        with pytest.raises(Http404):
+            self.request(user=user, url_kwargs={'project': 'abc123'})
+
+    def test_get_non_existend_pdfform(self):
```
Comment (bjohare) [inline cadasta/questionnaires/tests/test_views_default.py]: `test_get_non_existent_pdfform`

File: cadasta/questionnaires/tests/test_views_default.py [hunk 34]
```diff
@@ -0,0 +1,537 @@
...
+
+    def test_post_with_authorized_user(self):
+        user = UserFactory.create()
+        assign_policies(user)
+        response = self.request(method='POST', user=user)
+        assert response.status_code == 302
+        assert response.location == self.expected_success_url
+
+        # attributes field is deferred so we fetch a fresh instance
+        pdfform = PDFForm.objects.defer(None).get(id=self.pdfform.id)
```
Comment (bjohare) [inline cadasta/questionnaires/tests/test_views_default.py]: There are no attributes on PDFForm models so the call to defer is not required.

File: cadasta/questionnaires/tests/test_views_default.py [hunk 35]
```diff
@@ -0,0 +1,537 @@
...
+        assert pdfform.name == self.post_data['name']
+
+    def test_post_with_unauthorized_user(self):
+        user = UserFactory.create()
+        response = self.request(method='POST', user=user)
+        assert response.status_code == 302
+        assert ("You don't have permission to edit pdf form "
```
Comment (bjohare) [inline cadasta/questionnaires/tests/test_views_default.py]: `You don't have permission to edit this pdf form`

File: cadasta/questionnaires/tests/test_views_default.py [hunk 36]
```diff
@@ -0,0 +1,537 @@
...
+        response = self.request(user=self.user, method='POST')
+        assert response.status_code == 200
+        assert (response.headers['content-disposition'][1] ==
+                'attachment; filename={}.pdf'.format(self.pdfform.name))
+        assert (response.headers['content-type'][1] == 'application/pdf')
+
+    def test_post_with_unauthorized_user(self):
+        response = self.request(user=self.user, method='POST')
+        assert response.status_code == 302
+        assert ("You don't have permission to generate pdf forms"
```
Comment (bjohare) [inline cadasta/questionnaires/tests/test_views_default.py]: `You don't have permission to generate pdf forms for this project`

File: cadasta/questionnaires/views/default.py [hunk 37]
```diff
@@ -0,0 +1,91 @@
...
+from organization.views import mixins as organization_mixins
+from . import mixins
+from .. import messages as error_messages
+from core.mixins import LoginPermissionRequiredMixin
+from django.core.urlresolvers import reverse
+from django.http import HttpResponse
+from ..download.generator import PDFGenerator
+
+
+class PDFFormResources(LoginPermissionRequiredMixin,
```
Comment (bjohare) [inline cadasta/questionnaires/views/default.py]: This view should be renamed to `PDFFormList` as PDF forms are not project resources.

File: cadasta/questionnaires/views/default.py [hunk 38]
```diff
@@ -0,0 +1,91 @@
...
+    def get_perms_objects(self):
+        return [self.get_project()]
+
+    def get_queryset(self):
+        self.proj = self.get_project()
+        pdfforms = self.proj.pdfforms.all()
+        return pdfforms
+
+
+class PDFFormResourceNew(LoginPermissionRequiredMixin,
```
Comment (bjohare) [inline cadasta/questionnaires/views/default.py]: Should be renamed to `PDFFormAdd`

File: cadasta/questionnaires/views/default.py [hunk 39]
```diff
@@ -0,0 +1,91 @@
...
+    template_name = 'questionnaires/form_add_new.html'
+    permission_required = 'pdfform.add'
+    permission_denied_message = error_messages.QUESTIONNAIRE_PDF_FORM_ADD
+    model = PDFForm
+
+    def get_perms_objects(self):
+        return [self.get_project()]
+
+
+class PDFFormResourceDelete(LoginPermissionRequiredMixin,
```
Comment (bjohare) [inline cadasta/questionnaires/views/default.py]: rename to `PDFFormDelete`

File: cadasta/questionnaires/views/default.py [hunk 40]
```diff
@@ -0,0 +1,91 @@
...
+    permission_required = 'pdfform.delete'
+    permission_denied_message = error_messages.QUESTIONNAIRE_PDF_FORM_DELETE
+
+    def get_success_url(self):
+        kwargs = self.kwargs
+        del kwargs['form']
+        return reverse('questionnaires:pdf_form_list', kwargs=self.kwargs)
+
+
+class PDFFormResourceDetails(LoginPermissionRequiredMixin,
```
Comment (bjohare) [inline cadasta/questionnaires/views/default.py]: rename to `PDFFormDetail`

File: cadasta/questionnaires/views/default.py [hunk 41]
```diff
@@ -0,0 +1,91 @@
...
+                             mixins.PDFFormObjectMixin,
+                             organization_mixins.ProjectAdminCheckMixin,
+                             generic.DetailView):
+    template_name = 'questionnaires/form_detail.html'
+    permission_required = 'pdfform.view'
+    permission_denied_message = error_messages.QUESTIONNAIRE_PDF_FORM_VIEW
+    model = PDFForm
+
+
+class PDFFormResourcesEdit(LoginPermissionRequiredMixin,
```
Comment (bjohare) [inline cadasta/questionnaires/views/default.py]: rename to `PDFFormEdit`

File: cadasta/questionnaires/download/generator.py (hunk 2 above)
Comment (kavindya89) [inline cadasta/questionnaires/download/generator.py]: If we removed it there will be large amount of weasyprint warnings printed. I think we need to keep it.

Comment (bjohare) [inline cadasta/questionnaires/download/generator.py]: Ah ok that makes sense. I notice a warning about Pango when running `./runserver` is there a way to supress that?

```
WARNING: @font-face support needs Pango >= 1.38
```

**Comment by bjohare** [comment]
Closing. Will reopen as a new PR.

**Pull request closed without merging**